app/services/katello/ui_notifications/subscriptions/expire_soon.rb

```diff
@@ -0,0 +1,62 @@
+module Katello
+  module UINotifications
+    module Subscriptions
+      class ExpireSoon
+        class << self
+          def deliver!
+            subs_expiring_soon.each do |subscription|
+              next if notification_already_exists?(subscription) &&
+                update_expiration_date(subscription).zero?
+              ::Notification.create!(
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_SUBJECT,
+                :message => ::UINotifications::StringParser.new(
+                  blueprint.message,
+                  :subject => subscription
+                ),
+                :actions => {
+                  :links => [
+                    {
+                      # Can't use simply a :subscription_path on the blueprint
+                      # since this is a Bastion page
+                      :href => "/subscriptions/#{subscription.id}",
+                      :title => _('Subscription'),
+                      :external => true
+                    }
+                  ]
+                },
+                :notification_blueprint => blueprint
+              )
+            end
+          end
+
+          def subs_expiring_soon
+            Katello::Subscription.unscoped.all.select do |subscription|
+              subscription.expiring_soon?
+            end
+          end
+
+          def notification_already_exists?(subject)
+            subs_expiration_notification = Notification.unscoped.find_by(:subject => subject)
+            return false if subs_expiration_notification.blank?
+            subs_expiration_notification.notification_blueprint == blueprint
+          end
+
+          def update_expiration_date(subject)
+            expiration_date = subject.pools.
+              find { |pool| pool.expiring_soon? }.end_date.to_date
```

Comment:
this can return nil if none of the pools return true on the block (i.e. no pool is found), and this causes an exception (undefined method end_date for nil class)